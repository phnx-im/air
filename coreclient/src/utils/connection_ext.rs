// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use sqlx::{Connection, SqliteConnection, SqlitePool, SqliteTransaction};

use crate::store::StoreNotifier;

pub(crate) trait ConnectionExt {
    /// Executes a function with a transaction.
    ///
    /// The transaction is committed if the function returns `Ok`, and rolled
    /// back if the function returns `Err`.
    async fn with_transaction<T: Send>(
        &mut self,
        f: impl AsyncFnOnce(&mut SqliteTransaction<'_>) -> anyhow::Result<T>,
    ) -> anyhow::Result<T>;
}

impl ConnectionExt for SqliteConnection {
    async fn with_transaction<T: Send>(
        &mut self,
        f: impl AsyncFnOnce(&mut SqliteTransaction<'_>) -> anyhow::Result<T>,
    ) -> anyhow::Result<T> {
        let mut txn = self.begin_with("BEGIN IMMEDIATE").await?;
        let value = f(&mut txn).await?;
        txn.commit().await?;
        Ok(value)
    }
}

pub(crate) trait StoreExt {
    fn pool(&self) -> &SqlitePool;

    fn notifier(&self) -> StoreNotifier;

    /// Executes a function with a transaction.
    ///
    /// The transaction is committed if the function returns `Ok`, and rolled
    /// back if the function returns `Err`.
    async fn with_transaction<U: Send>(
        &self,
        f: impl AsyncFnOnce(&mut SqliteTransaction<'_>) -> anyhow::Result<U>,
    ) -> anyhow::Result<U> {
        let mut txn = self.pool().begin_with("BEGIN IMMEDIATE").await?;
        let value = f(&mut txn).await?;
        txn.commit().await?;
        Ok(value)
    }

    /// Executes a function with a [`StoreNotifier`].
    ///
    /// The [`StoreNotifier`] is notified if the function returns `Ok`.
    async fn with_notifier<T: Send>(
        &self,
        f: impl AsyncFnOnce(&mut StoreNotifier) -> anyhow::Result<T>,
    ) -> anyhow::Result<T> {
        let mut notifier = self.notifier();
        let value = f(&mut notifier).await?;
        notifier.notify();
        Ok(value)
    }

    /// Executes a function with a transaction and a [`StoreNotifier`].
    ///
    /// The transaction is committed if the function returns `Ok`, and rolled
    /// back if the function returns `Err`. The [`StoreNotifier`] is notified
    /// after the transaction is committed successfully.
    async fn with_transaction_and_notifier<T: Send>(
        &self,
        f: impl AsyncFnOnce(&mut SqliteTransaction<'_>, &mut StoreNotifier) -> anyhow::Result<T>,
    ) -> anyhow::Result<T> {
        let mut txn = self.pool().begin_with("BEGIN IMMEDIATE").await?;
        let mut notifier = self.notifier();
        let value = f(&mut txn, &mut notifier).await?;
        txn.commit().await?;
        notifier.notify();
        Ok(value)
    }
}
