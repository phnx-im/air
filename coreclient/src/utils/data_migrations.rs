// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Data migrations implemented in Rust that cannot be expressed in SQL.

use aircommon::codec::{AirCodec, BlobDecoded, BlobEncoded};
use mimi_content::content_container::MimiContentV1;
use sqlx::{SqlitePool, migrate::Migrate, query, query_as};
use tokio_stream::StreamExt;
use tracing::{error, info};

use crate::{ConversationMessageId, conversations::messages::persistence::VersionedMessage};

const MESSAGE_STATUS_MIGRATION_VERSION: i64 = 20250703133517;

/// Migrate data in the database that cannot be expressed in SQL.
pub(crate) async fn migrate(pool: &SqlitePool) -> Result<(), sqlx::Error> {
    let Some(migrations) = pool.acquire().await?.list_applied_migrations().await.ok() else {
        // The migrations might not yet exist
        return Ok(());
    };

    let has_message_status = migrations
        .iter()
        .any(|m| m.version == MESSAGE_STATUS_MIGRATION_VERSION);
    if !has_message_status && let Err(error) = convert_messages_v1_to_v2(pool).await {
        error!(%error, "Failed to convert messages from version 1 to version 2");
    }

    Ok(())
}

/// Convert all versioned v1 messages to v2 messages in the database.
async fn convert_messages_v1_to_v2(pool: &SqlitePool) -> anyhow::Result<usize> {
    info!("Data migration: Converting messages from version 1 to version 2");

    let mut write_connection = pool.acquire().await?;
    let mut read_connection = pool.acquire().await?;

    struct Record {
        message_id: ConversationMessageId,
        content: BlobDecoded<VersionedMessage>,
    }

    let mut records = query_as!(
        Record,
        r#"SELECT
            message_id AS "message_id: _",
            content AS "content: _"
        FROM conversation_messages"#
    )
    .fetch(read_connection.as_mut());

    let mut num_messages = 0;

    while let Some(Record {
        message_id,
        content: BlobDecoded(VersionedMessage { version, content }),
    }) = records.next().await.transpose()?
        && version == 1
    {
        let content_v1: MimiContentV1 = AirCodec::from_slice(&content)?;
        let content_v2 = content_v1.upgrade();
        let Ok(message) = VersionedMessage::from_mimi_content(&content_v2) else {
            error!(
                ?message_id,
                "Failed to convert message from version 1 to version 2; skip"
            );
            continue;
        };
        let message = BlobEncoded(message);
        query!(
            "UPDATE conversation_messages SET content = ? WHERE message_id = ?",
            message,
            message_id,
        )
        .execute(&mut *write_connection)
        .await?;

        num_messages += 1;
    }

    info!(
        num_messages,
        "Converted messages from version 1 to version 2",
    );

    Ok(num_messages)
}

#[cfg(test)]
mod test {
    use aircommon::{assert_matches, identifiers::UserId, time::TimeStamp};
    use sqlx::{SqliteConnection, migrate::Migrate};

    use crate::{
        ConversationId,
        conversations::{
            messages::persistence::tests::test_conversation_message,
            persistence::tests::test_conversation,
        },
        store::StoreNotifier,
    };

    use super::*;

    async fn store_raw_test_message_v1(
        connection: &mut SqliteConnection,
        converation_id: ConversationId,
    ) -> anyhow::Result<()> {
        let message_id = ConversationMessageId::random();
        let user_id = UserId::random("localhost".parse().unwrap());
        let user_uuid = user_id.uuid();
        let user_domain = user_id.domain();
        let timestamp = TimeStamp::now();
        let mimi_content = MimiContentV1 {
            topic_id: vec![1; 32].into(),
            ..Default::default()
        };
        let mimi_content_bytes = AirCodec::to_vec(&mimi_content).unwrap();
        let content = VersionedMessage {
            version: 1,
            content: mimi_content_bytes,
        };

        sqlx::query(
            "INSERT INTO conversation_messages (
                message_id,
                conversation_id,
                sender_user_uuid,
                sender_user_domain,
                timestamp,
                content,
                sent
            ) VALUES (?, ?, ?, ?, ?, ?, ?)",
        )
        .bind(message_id)
        .bind(converation_id)
        .bind(user_uuid)
        .bind(user_domain)
        .bind(timestamp)
        .bind(BlobEncoded(content))
        .bind(false)
        .execute(connection)
        .await?;

        Ok(())
    }

    #[sqlx::test(migrations = false)]
    async fn convert_messages_v1_to_v2_some_messages(pool: SqlitePool) -> anyhow::Result<()> {
        let mut connection = pool.acquire().await?;
        connection.ensure_migrations_table().await?;
        let migrator = sqlx::migrate!();

        let itx = migrator
            .iter()
            .position(|m| m.version == MESSAGE_STATUS_MIGRATION_VERSION)
            .unwrap();

        // apply migrations up to the message status migration
        for migration in migrator.iter().take(itx) {
            connection.apply(migration).await?;
        }

        let mut notifier = StoreNotifier::noop();
        let conversation = test_conversation();
        conversation.store(&mut connection, &mut notifier).await?;

        // Store v1 messages
        store_raw_test_message_v1(&mut connection, conversation.id()).await?;
        store_raw_test_message_v1(&mut connection, conversation.id()).await?;
        store_raw_test_message_v1(&mut connection, conversation.id()).await?;

        // finish migration
        for migration in migrator.iter().skip(itx) {
            connection.apply(migration).await?;
        }

        // Store v2 messages
        test_conversation_message(conversation.id())
            .store(connection.as_mut(), &mut notifier)
            .await?;
        test_conversation_message(conversation.id())
            .store(connection.as_mut(), &mut notifier)
            .await?;
        test_conversation_message(conversation.id())
            .store(connection.as_mut(), &mut notifier)
            .await?;

        assert_matches!(convert_messages_v1_to_v2(&pool).await, Ok(3));

        Ok(())
    }
}
