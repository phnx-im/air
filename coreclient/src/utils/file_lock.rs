// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::{
    fs::{File, OpenOptions},
    io,
    path::Path,
    sync::Arc,
};

use tokio::task::spawn_blocking;

#[derive(Debug)]
pub(crate) struct FileLock {
    file: Arc<File>,
}

impl FileLock {
    pub(crate) fn new(path: impl AsRef<Path>) -> io::Result<Self> {
        let file = OpenOptions::new()
            .write(true)
            .read(true)
            .create(true)
            .truncate(true)
            .open(path)?;
        Ok(Self {
            file: Arc::new(file),
        })
    }

    pub(crate) fn from_file(file: File) -> io::Result<Self> {
        Ok(Self {
            file: Arc::new(file),
        })
    }

    /// Note: `&mut self` makes sure that the file cannot be locked twice which is unspecified
    /// behavior and platform dependent.
    pub(crate) async fn lock(&mut self) -> io::Result<FileLockGuard<'_>> {
        let file = self.file.clone();
        spawn_blocking(move || file.lock())
            .await
            .map_err(|_| io::Error::other("background task failed"))??;
        Ok(FileLockGuard {
            file_lock: Some(self),
        })
    }
}

#[derive(Debug)]
#[must_use]
pub struct FileLockGuard<'a> {
    file_lock: Option<&'a mut FileLock>,
}

impl Drop for FileLockGuard<'_> {
    fn drop(&mut self) {
        if let Some(lock_file) = self.file_lock.take() {
            let _ = lock_file.file.unlock();
        }
    }
}

#[cfg(test)]
mod tests {
    use tempfile::tempdir;
    use tokio::time::{Duration, timeout};

    use super::*;

    #[tokio::test]
    async fn lock_and_unlock() -> anyhow::Result<()> {
        let dir = tempdir().unwrap();
        let path = dir.path().join("lockfile");

        let mut file_lock = FileLock::new(&path)?;
        let guard = file_lock.lock().await.expect("failed to lock file");

        let mut file_lock2 = FileLock::new(&path)?;
        timeout(Duration::from_millis(100), file_lock2.lock())
            .await
            .expect_err("another file lock should timeout");

        drop(guard);

        let _guard = file_lock2.lock().await?;

        Ok(())
    }

    #[tokio::test]
    async fn blocks_until_unlocked() -> anyhow::Result<()> {
        let dir = tempdir().unwrap();
        let path = dir.path().join("lockfile");

        let mut file_lock = FileLock::new(&path)?;
        let guard = file_lock.lock().await.expect("failed to lock file");

        let handle = tokio::spawn(async move {
            let mut file_lock = FileLock::new(&path)?;
            file_lock.lock().await?;
            Ok::<(), anyhow::Error>(())
        });

        drop(guard);
        handle.await??;

        Ok(())
    }
}
