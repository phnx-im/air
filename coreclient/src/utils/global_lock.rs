#[cfg(any(test, feature = "test_utils"))]
use std::fs::File;
use std::{io, path::Path};

#[derive(Debug)]
pub(crate) struct GlobalLock {
    #[cfg(not(target_os = "android"))]
    file: super::file_lock::FileLock,
    #[cfg(target_os = "android")]
    lock: tokio::sync::Mutex<()>,
}

impl GlobalLock {
    pub(crate) fn new(_path: impl AsRef<Path>) -> io::Result<Self> {
        #[cfg(not(target_os = "android"))]
        {
            Ok(Self {
                file: super::file_lock::FileLock::new(_path)?,
            })
        }
        #[cfg(target_os = "android")]
        {
            Ok(Self {
                lock: tokio::sync::Mutex::new(()),
            })
        }
    }

    #[cfg(any(test, feature = "test_utils"))]
    pub(crate) fn from_file(_file: File) -> Self {
        #[cfg(not(target_os = "android"))]
        {
            Self {
                file: super::file_lock::FileLock::from_file(_file),
            }
        }
        #[cfg(target_os = "android")]
        {
            Self {
                lock: tokio::sync::Mutex::new(()),
            }
        }
    }

    /// Note: `&mut self` makes sure that the file cannot be locked twice which is unspecified
    /// behavior and platform dependent.
    pub(crate) async fn lock<'a>(&'a mut self) -> io::Result<impl Drop + 'a> {
        #[cfg(not(target_os = "android"))]
        {
            self.file.lock().await
        }
        #[cfg(target_os = "android")]
        {
            Ok(self.lock.lock().await)
        }
    }
}
