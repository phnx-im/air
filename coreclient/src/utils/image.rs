// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::io::Cursor;

use exif::{Exif, Tag};
use image::{DynamicImage, GenericImageView};
use tracing::info;

pub(crate) fn resize_profile_image(mut image_bytes: &[u8]) -> anyhow::Result<Vec<u8>> {
    let image = image::load_from_memory(image_bytes)?;

    // Read EXIF data
    let exif_reader = exif::Reader::new();
    let mut image_bytes_cursor = Cursor::new(&mut image_bytes);
    let exif = exif_reader
        .read_from_container(&mut image_bytes_cursor)
        .ok();

    // Resize the image
    let image = image.resize(256, 256, image::imageops::FilterType::Nearest);

    let image = rotate(exif, image);

    // Save the resized image
    let mut buf = Vec::new();
    let mut cursor = Cursor::new(&mut buf);
    let mut encoder = image::codecs::jpeg::JpegEncoder::new_with_quality(&mut cursor, 90);
    encoder.encode_image(&image)?;
    info!(
        from_bytes = image_bytes.len(),
        to_bytes = buf.len(),
        "Resized profile image",
    );
    Ok(buf)
}

const ATTACHMENT_IMAGE_QUALITY_PERCENT: f32 = 90.0;
const MAX_ATTACHMENT_IMAGE_WIDTH: u32 = 4096;
const MAX_ATTACHMENT_IMAGE_HEIGHT: u32 = 4096;
const ATTACHMENT_THUMBNAIL_WIDTH: u32 = 300;
const ATTACHMENT_THUMBNAIL_HEIGHT: u32 = 300;

pub(crate) struct ReencodedAttachmentImage {
    pub(crate) webp_image: Vec<u8>,
    pub(crate) image_dimensions: (u32, u32),
    pub(crate) webp_thumbnail: Vec<u8>,
    pub(crate) blurhash: String,
}

/// Reencodes the image to WEBP format.
///
/// This does several things:
/// - Rotates and flips the image according to the EXIF orientation
/// - Resizes the image to a maximum width and height of 4096x4096
/// - Converts the image to WebP
///
/// Returns the WebP image bytes and the blurhash of the image.
pub(crate) fn reencode_attachment_image(
    image_bytes: Vec<u8>,
) -> anyhow::Result<ReencodedAttachmentImage> {
    let mut image_bytes = image_bytes.as_slice();
    let image = image::load_from_memory(image_bytes)?;

    // Read EXIF data
    let exif_reader = exif::Reader::new();
    let mut image_bytes_cursor = Cursor::new(&mut image_bytes);
    let exif = exif_reader
        .read_from_container(&mut image_bytes_cursor)
        .ok();

    let image = rotate(exif, image);
    let image = resize(image);

    // TODO: Preserve format instead of converting to WebP

    let image_rgba = image.to_rgba8();
    let (width, height) = image_rgba.dimensions();

    let webp_image = webp::Encoder::from_rgba(&image_rgba, width, height)
        .encode(ATTACHMENT_IMAGE_QUALITY_PERCENT);

    // `blurhash::encode` can only fail if the compoments dimension is out of range
    // => We should never get an error here.
    let blurhash = blurhash::encode(4, 3, width, height, &image_rgba)?;

    let thumbnail = image
        .thumbnail(ATTACHMENT_THUMBNAIL_WIDTH, ATTACHMENT_THUMBNAIL_HEIGHT)
        .to_rgba8();
    let (thumbnail_width, thumbnail_height) = thumbnail.dimensions();
    let webp_thumbnail = webp::Encoder::from_rgba(&thumbnail, thumbnail_width, thumbnail_height)
        .encode(ATTACHMENT_IMAGE_QUALITY_PERCENT);

    info!(
        from_bytes = image_bytes.len(),
        to_bytes = webp_image.len(),
        "Reencoded attachment image as WebP",
    );

    // Note: We need to convert WebPMemory to Vec here, because the former is not Send.
    Ok(ReencodedAttachmentImage {
        webp_image: webp_image.to_vec(),
        image_dimensions: (width, height),
        webp_thumbnail: webp_thumbnail.to_vec(),
        blurhash,
    })
}

// Rotate/flip the image according to the orientation if necessary
fn rotate(exif: Option<Exif>, image: DynamicImage) -> DynamicImage {
    if let Some(exif) = exif {
        let orientation = exif
            .get_field(Tag::Orientation, exif::In::PRIMARY)
            .and_then(|field| field.value.get_uint(0))
            .unwrap_or(1);
        // TODO: roate and flip in-place
        match orientation {
            1 => image,
            2 => image.fliph(),
            3 => image.rotate180(),
            4 => image.flipv(),
            5 => image.rotate90().fliph(),
            6 => image.rotate90(),
            7 => image.rotate270().fliph(),
            8 => image.rotate270(),
            _ => image,
        }
    } else {
        image
    }
}

fn resize(image: DynamicImage) -> DynamicImage {
    let (width, height) = image.dimensions();

    if width <= MAX_ATTACHMENT_IMAGE_WIDTH && height <= MAX_ATTACHMENT_IMAGE_HEIGHT {
        return image;
    }

    let scale_x = MAX_ATTACHMENT_IMAGE_WIDTH as f32 / width as f32;
    let scale_y = MAX_ATTACHMENT_IMAGE_HEIGHT as f32 / height as f32;
    let scale = scale_x.min(scale_y);

    let new_width = (width as f32 * scale).round() as u32;
    let new_height = (height as f32 * scale).round() as u32;

    image.resize(new_width, new_height, image::imageops::FilterType::Lanczos3)
}
