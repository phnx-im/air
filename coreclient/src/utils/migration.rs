// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use migrations::EmbeddedMigration;
use refinery::Migration;

refinery::embed_migrations!("migrations");

pub(crate) fn run_migrations(
    client_db_connection: &mut rusqlite::Connection,
) -> Result<(), refinery::Error> {
    for migration in migrations::runner().run_iter(client_db_connection) {
        post_process(migration?);
    }

    match migrations::runner().run(client_db_connection) {
        Ok(report) => {
            log::info!(
                "Applied migrations successfully. Migrations applied: {}",
                report.applied_migrations().len()
            );
            Ok(())
        }
        Err(e) => {
            log::error!("Failed to apply migrations: {}", e);
            Err(e)
        }
    }
}

fn post_process(migration: Migration) {
    match migration.into() {
        EmbeddedMigration::CreateInitialTablesAndTriggers(_) => {
            // Perform post-processing for arbitrary migrations here.
        }
    }
}
