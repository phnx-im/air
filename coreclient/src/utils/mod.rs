// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

pub(crate) mod connection_ext;
pub(crate) mod data_migrations;
pub(crate) mod image;
pub(crate) mod persistence;

#[cfg(test)]
pub(crate) fn init_test_tracing() {
    use tracing::Level;
    use tracing_subscriber::EnvFilter;

    let _ = tracing_subscriber::fmt::fmt()
        .with_test_writer()
        .with_env_filter(
            EnvFilter::builder()
                .with_default_directive(Level::INFO.into())
                .from_env_lossy(),
        )
        .try_init();
}
