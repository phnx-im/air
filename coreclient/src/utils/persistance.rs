// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use phnxbackend::auth_service::AsClientId;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use tls_codec::Serialize as TlsSerializeTrait;
use turbosql::{execute, select, Turbosql};

#[derive(Turbosql)]
struct TurboData {
    rowid: Option<i64>,
    data_type: Option<Vec<u8>>,
    client_id: Option<Vec<u8>>,
    key: Option<Vec<u8>>,
    secondary_key: Option<Vec<u8>>,
    value: Option<Vec<u8>>,
}

impl TurboData {
    fn from_persistable<T: Persistable>(value: &T) -> Result<Self, turbosql::Error> {
        let client_id_bytes = value.own_client_id_bytes();
        let key_bytes = serde_json::to_vec(&value.key())?;
        let secondary_key_bytes = serde_json::to_vec(&value.secondary_key())?;
        let data_type_bytes = serde_json::to_vec(&T::DATA_TYPE)?;
        let value_bytes = serde_json::to_vec(&value)?;
        Ok(Self {
            rowid: value.rowid(),
            data_type: Some(data_type_bytes),
            client_id: Some(client_id_bytes),
            key: Some(key_bytes),
            value: Some(value_bytes),
            secondary_key: Some(secondary_key_bytes),
        })
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub(crate) enum DataType {
    Contact,
    Conversation,
    Group,
    Message,
    AsCredential,
}

pub(crate) trait Persistable: Serialize + DeserializeOwned {
    type Key: Serialize + std::fmt::Debug;
    type SecondaryKey: Serialize + std::fmt::Debug;

    const DATA_TYPE: DataType;

    fn own_client_id_bytes(&self) -> Vec<u8>;

    fn rowid(&self) -> Option<i64>;

    fn key(&self) -> &Self::Key;

    fn secondary_key(&self) -> &Self::SecondaryKey;

    fn load(own_client_id: &AsClientId, key: &Self::Key) -> Result<Self, turbosql::Error> {
        let client_id_bytes = own_client_id.tls_serialize_detached().unwrap();
        let key_bytes = serde_json::to_vec(key)?;
        let data_type_bytes = serde_json::to_vec(&Self::DATA_TYPE)?;
        let turbo_data = select!(TurboData "WHERE client_id = " client_id_bytes " AND key = " key_bytes " AND data_type = " data_type_bytes)?;
        let value = turbo_data
            .value
            .ok_or(turbosql::Error::OtherError("Could not load value from DB."))?;
        Ok(serde_json::from_slice(&value)?)
    }

    fn load_secondary(
        own_client_id: &AsClientId,
        secondary_key: &Self::SecondaryKey,
    ) -> Result<Self, turbosql::Error> {
        let client_id_bytes = own_client_id.tls_serialize_detached().unwrap();
        let key_bytes = serde_json::to_vec(secondary_key)?;
        let data_type_bytes = serde_json::to_vec(&Self::DATA_TYPE)?;
        let turbo_data = select!(TurboData "WHERE client_id = " client_id_bytes " AND secondary_key = " key_bytes " AND data_type = " data_type_bytes)?;
        let value = turbo_data
            .value
            .ok_or(turbosql::Error::OtherError("Could not load value from DB."))?;
        Ok(serde_json::from_slice(&value)?)
    }

    fn load_multiple_secondary(
        own_client_id: &AsClientId,
        secondary_key: &Self::SecondaryKey,
    ) -> Result<Vec<Self>, turbosql::Error> {
        let client_id_bytes = own_client_id.tls_serialize_detached().unwrap();
        let data_type_bytes = serde_json::to_vec(&Self::DATA_TYPE)?;
        let key_bytes = serde_json::to_vec(secondary_key)?;
        let values = select!(Vec<TurboData> "WHERE client_id = " client_id_bytes " AND secondary_key = " key_bytes " AND data_type = " data_type_bytes)?;
        let mapped_values = values
            .into_iter()
            .map(|turbo_data| {
                let value = turbo_data
                    .value
                    .ok_or(turbosql::Error::OtherError("Could not load value from DB."))?;
                Ok(serde_json::from_slice(&value)?)
            })
            .collect::<Result<Vec<_>, turbosql::Error>>()?;
        Ok(mapped_values)
    }

    fn load_all(own_client_id: &AsClientId) -> Result<Vec<Self>, turbosql::Error> {
        let client_id_bytes = own_client_id.tls_serialize_detached().unwrap();
        let data_type_bytes = serde_json::to_vec(&Self::DATA_TYPE)?;
        let values = select!(Vec<TurboData> "WHERE client_id = " client_id_bytes " AND data_type = " data_type_bytes)?;
        let mapped_values = values
            .into_iter()
            .map(|turbo_data| {
                let value = turbo_data
                    .value
                    .ok_or(turbosql::Error::OtherError("Could not load value from DB."))?;
                Ok(serde_json::from_slice(&value)?)
            })
            .collect::<Result<Vec<_>, turbosql::Error>>()?;
        Ok(mapped_values)
    }

    fn persist(&self) -> Result<(), turbosql::Error> {
        let turbo_data = TurboData::from_persistable(self)?;
        if self.rowid().is_some() {
            turbo_data.update()?;
        } else {
            // We can unwrap these, as they are both set in the constructor.
            let client_id_bytes = turbo_data.client_id.as_ref().unwrap();
            let key_bytes = turbo_data.key.as_ref().unwrap();
            let data_type_bytes = serde_json::to_vec(&Self::DATA_TYPE)?;
            // Check if a contact with this ID already exists.
            if let Ok(old_data) = select!(TurboData "WHERE client_id = " client_id_bytes " AND key = " key_bytes " AND data_type = " data_type_bytes)
            {
                // If it exists, delete it from the DB. (We could probably just
                // read out the rowid of the existing contact and set it for the
                // new contact, but this does the trick.)
                execute!("DELETE FROM turbodata WHERE rowid = " old_data.rowid.unwrap())?;
            }
            // Insert the new data into the DB.
            turbo_data.insert()?;
        }
        Ok(())
    }

    fn _purge(&self) -> Result<(), turbosql::Error> {
        let turbo_data = TurboData::from_persistable(self)?;
        let rowid = turbo_data.rowid.ok_or(turbosql::Error::OtherError(
            "Cannot purge data without rowid.",
        ))?;
        execute!("DELETE FROM turbodata WHERE rowid = " rowid)?;
        Ok(())
    }
}
