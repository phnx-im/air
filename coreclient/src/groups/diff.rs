// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::collections::BTreeMap;

use phnx_types::credentials::{keys::InfraCredentialSigningKey, ClientCredential};

use crate::utils::{deserialize_btreemap, serialize_hashmap};

use super::*;

/// A struct that contains differences in group data when creating a commit.
/// The diff of a group should be merged when the pending commit of the
/// underlying MLS group is merged.
#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct GroupDiff {
    pub(crate) leaf_signer: Option<InfraCredentialSigningKey>,
    pub(crate) signature_ear_key: Option<SignatureEarKeyWrapperKey>,
    pub(crate) credential_ear_key: Option<ClientCredentialEarKey>,
    pub(crate) group_state_ear_key: Option<GroupStateEarKey>,
    pub(crate) user_auth_key: Option<UserAuthSigningKey>,
    // Changes to the client credentials. `None` denotes a deleted credential at
    // the given index, `Some` denotes an added or updated credential. The
    // vector must be sorted in ascending order of indices.
    #[serde(
        serialize_with = "serialize_hashmap",
        deserialize_with = "deserialize_btreemap"
    )]
    pub(crate) client_information: BTreeMap<usize, Option<(ClientCredential, SignatureEarKey)>>,
    pub(crate) new_number_of_leaves: usize,
}

impl GroupDiff {
    pub(crate) fn new(group: &Group) -> Self {
        Self {
            leaf_signer: None,
            signature_ear_key: None,
            credential_ear_key: None,
            group_state_ear_key: None,
            user_auth_key: None,
            client_information: BTreeMap::new(),
            new_number_of_leaves: group
                .client_information
                .last_key_value()
                .map(|(index, _)| index + 1)
                .unwrap_or(0),
        }
    }

    /// This overrides any previous changes to the client credentials.
    pub(crate) fn remove_client_credential(&mut self, removed_index: LeafNodeIndex) {
        self.client_information.insert(removed_index.usize(), None);
    }

    pub(crate) fn client_information<'a>(
        &'a self,
        index: usize,
        existing_client_information: &'a BTreeMap<usize, (ClientCredential, SignatureEarKey)>,
    ) -> Option<&'a (ClientCredential, SignatureEarKey)> {
        if let Some(Some(credential)) = self.client_information.get(&index) {
            Some(credential)
        } else {
            existing_client_information.get(&index)
        }
    }

    /// Add a client credential in the first free index, or extend the current
    /// list of credentials. This takes into account previous removes.
    pub(crate) fn add_client_information(
        &mut self,
        existing_client_information: &BTreeMap<usize, (ClientCredential, SignatureEarKey)>,
        new_client_information: (ClientCredential, SignatureEarKey),
    ) {
        for index in 0..self.new_number_of_leaves {
            if self
                .client_information(index, existing_client_information)
                .is_none()
            {
                self.client_information
                    .insert(index, Some(new_client_information));
                return;
            }
        }
        // If we're still here, we have not found a free index yet and we have
        // to extend the vector of credentials.
        self.client_information
            .insert(self.new_number_of_leaves, Some(new_client_information));
        self.new_number_of_leaves += 1;
    }

    pub(crate) fn apply_pending_removes(&mut self, staged_commit: &StagedCommit) {
        for pending_remove in staged_commit.remove_proposals() {
            self.remove_client_credential(pending_remove.remove_proposal().removed())
        }
    }
}
