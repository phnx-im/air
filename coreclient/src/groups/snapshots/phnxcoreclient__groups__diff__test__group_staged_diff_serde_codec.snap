---
source: coreclient/src/groups/diff.rs
expression: "PhnxCodec::to_vec(&*STAGED_GROUP_DIFF).unwrap()"
extension: cbor
snapshot_kind: binary
---
